File: tools/hexlifyscript.py
#!/usr/bin/env python3

'''
Turn a Python script into Intel HEX format to be concatenated at the
end of the MicroPython firmware.hex.  A simple header is added to the
script.

To execute from command line: ./hexlifyscript.py <script.py>
It also accepts data on standard input.
'''

import struct
import binascii
import fileinput


IHEX_TYPE_DATA = 0
IHEX_TYPE_EXT_LIN_ADDR = 4

SCRIPT_ADDR = 0x3e000 # magic start address in flash of script

def make_ihex_record(addr, type, data):
    record = struct.pack('>BHB', len(data), addr & 0xffff, type) + data
    checksum = (-(sum(record))) & 0xff
    return ':%s%02X' % (str(binascii.hexlify(record), 'utf8').upper(), checksum)

def hexlify_script(script):
    # add header, pad to multiple of 16 bytes
    data = b'MP' + struct.pack('<H', len(script)) + script
    data = data + bytes(16 - len(data) % 16)
    assert len(data) <= 0x2000

    # convert to .hex format
    output = []
    addr = SCRIPT_ADDR
    assert(SCRIPT_ADDR >> 16 == 3) # 0x0003 is hard coded in line below
    output.append(make_ihex_record(0, IHEX_TYPE_EXT_LIN_ADDR, b'\x00\x03'))
    for i in range(0, len(data), 16):
        chunk = data[i:min(i + 16, len(data))]
        output.append(make_ihex_record(addr, IHEX_TYPE_DATA, chunk))
        addr += 16

    return '\n'.join(output)

if __name__ == '__main__':
    # read script from a file and print out the hexlified version
    with fileinput.input(mode='rb') as lines:
        print(hexlify_script(b''.join(lines)))

File: tools/hexlifyscript.js
/*
Turn a Python script into Intel HEX format to be concatenated at the
end of the MicroPython firmware.hex.  A simple header is added to the
script.

To execute from command line: node hexlifyscript.js <script.py>
*/

// hexlifyScript:
// - takes a Python script as a string
// - returns hexlified string, with newlines between lines
function hexlifyScript(script) {
    function hexlify(ar) {
        var result = '';
        for (var i = 0; i < ar.length; ++i) {
            if (ar[i] < 16) {
                result += '0';
            }
            result += ar[i].toString(16);
        }
        return result;
    }

    // add header, pad to multiple of 16 bytes
    data = new Uint8Array(4 + script.length + (16 - (4 + script.length) % 16));
    data[0] = 77; // 'M'
    data[1] = 80; // 'P'
    data[2] = script.length & 0xff;
    data[3] = (script.length >> 8) & 0xff;
    for (var i = 0; i < script.length; ++i) {
        data[4 + i] = script.charCodeAt(i);
    }
    // TODO check data.length < 0x2000

    // convert to .hex format
    var addr = 0x3e000; // magic start address in flash
    var chunk = new Uint8Array(5 + 16);
    var output = [];
    output.push(':020000040003F7') // extended linear address, 0x0003
    for (var i = 0; i < data.length; i += 16, addr += 16) {
        chunk[0] = 16; // length of data section
        chunk[1] = (addr >> 8) & 0xff; // high byte of 16-bit addr
        chunk[2] = addr & 0xff; // low byte of 16-bit addr
        chunk[3] = 0; // type (data)
        for (var j = 0; j < 16; ++j) {
            chunk[4 + j] = data[i + j];
        }
        var checksum = 0;
        for (var j = 0; j < 4 + 16; ++j) {
            checksum += chunk[j];
        }
        chunk[4 + 16] = (-checksum) & 0xff;
        output.push(':' + hexlify(chunk).toUpperCase())
    }

    return output.join('\n');
}

// read script from file, hexlify, then print to console
console.log(hexlifyScript(require('fs').readFileSync(process.argv[2], 'utf8')));
